src/v/ssx/sharded_ptr.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "base/vassert.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/sharded.hh>
+
+#include <memory>
+
+namespace ssx {
+
+/// A pointer that is shared between shards.
+/// reset() must be called on the home shard.
+/// The pointer is stable until the fiber yields.
+template<typename T>
+class sharded_ptr {
```

Comment:
The `shared_token_bucket` is `const`, due to a `const limit` (this probably makes the atomic operations manageable by limiting the number of atomic variables).

So I hold it by pointer. When the throughput limit is updated, I need to reseat that pointer with a new bucket that has a new limit. That must be done on the home core, but even with an atomic pointer, I can't guarantee that the pointed to bucket is stable between scheduling points.

We make multiple accesses to the token bucket, e.g.:
```c++
tb.duration_for(tb.deficiency(tb.grab(0))
```
So I need to pointer to be stable at least between scheduling points.

That's what `sharded_ptr` is for; it keeps the original alive, updates the local copy on other shards, and then takes ownership of the new pointer, and destroys the old one.

Reply:
Exactly.

Reply:
It would be nice to see a bit more detail on how it is to be used safely. For example, unlike most sharded services, some methods are safe to call concurrently from other shards even before `start()` (which happens implicitly) is called, right?

How about `stop()`: what are the semantics of concurrent access on other shards while `stop()` is called on the owner? Concurrent `stop()`s on the owner? Etc.

If possible we should try to enforce those rules, as you've done with the owning shard assert, since these are things that are easy to mis-use and then they seem to work until one day ... boom.

Reply:
Coming back to our shared ptr discussion and having thought about this some more.

Couldn't we just use it as follows:
 - Construct a `std::shared_ptr<ss::internal::token_bucket>` on init and a new one when updated
 - Copy and send to other shards via `submit_to` (or some `invoke_on` service wrapper)
 - Optionally: Take a copy at the start of each request

The former two should preserve the property that eventually all shards converge to the same value and no lifetime issues across shards. 

The last step would even guarantee consistency within a shard across suspension points if we want to go that far.

This would seem simpler to me. Thoughts?

Reply:
> that eventually all shards converge to the same value

I think you still need a mutex in there for this to be true? I.e., when a reset occurs you want the whole thing to occur, including the `submit_to` resolving on all shards before you allow the next reset to start.

I like this solution though: simpler, no raw pointers, and allows the other side to use the usual smart pointer semantics (including coping to a local as you mention), rather than the much more subtle rule of "reference only valid until next suspend".

Reply:
> I think you still need a mutex in there for this to be true?

Right yeah, guess it depends on a few things as you write in a different comment further down but doesn't hurt adding either way.

Reply:
> I like this solution though: simpler, no raw pointers, and allows the other side to use the usual smart pointer semantics (including coping to a local as you mention), rather than the much more subtle rule of "reference only valid until next suspend".

I'm not sure. I t would make this existing and seemingly valid code into a footgun since the bucket could be replaced by another thread half way through the statement.
```c++
delay_t eval_node_delay(
  const ssx::sharded_ptr<snc_quota_manager::bucket_t>& tbp) noexcept {
    if (!tbp) {
        return delay_t::zero();
    }
    auto& tb = *tbp;
    return std::chrono::duration_cast<delay_t>(
      tb.duration_for(tb.deficiency(tb.grab(0))));
}
```
Which doesn't seem harder to misuse than holding across an async point.

Reply:
Done. With `mutex` and std::vector<std::shared_ptr<T>>`

Reply:
AFAICT `std::shared_ptr<T>` instances never escape the interface (e.g. dereference operators proxy to the underlying pointer), so I'm not entirely sure why `_state` is a vector of `std::shared_ptr<T>` instances. That is, we'd pay an atomic increment cost to copy the instances into `_state`, but further accesses via the dereference operators aren't bumping the reference count. am i missing something here?

Reply:
> AFAICT `std::shared_ptr<T>` instances never escape the interface (e.g. dereference operators proxy to the underlying pointer), so I'm not entirely sure why `_state` is a vector of `std::shared_ptr<T>` instances. That is, we'd pay an atomic increment cost to copy the instances into `_state`, but further accesses via the dereference operators aren't bumping the reference count. am i missing something here?

It made the implementation a lot ccleaner.

But the intention was to allow a `shared_ptr` or `weak_ptr` to be held, although it's currently a bit YAGNI. Would making `local` public suffice, or would another name be more appropriate? `operator()` is available.

Reply:
> Would making local public suffice, or would another name be more appropriate? operator() is available.

suffice for what?

Reply:
A name that can be used like:
```c++
    sharded_ptr<int> sp;
    sp.reset(42);
    std::shared_ptr<int> shared = p0.local();
```

Reply:
I don't have a preference. I only was making the observation that if the shared pointer is not escaping the interface, then it didn't seem to serve a purpose to have one per core.

Reply:
@dotnwat - you need the vector of `shared_pointers` (or some other approach) for thread-safety, even if they don't escape the interface. With a single pointer shard 1 can be accessing the shared pointer, while shard 0 (the owning shard) is resetting it: this is not safe (e.g., the object could be destroyed after the pointer was read but before subsequent use, a UAF).

So the vector of shared pointers is to allow each shard to own a copy and update it in the usual message passing way.